core/metacling/src/TCling.cxx

```diff
@@ -5891,11 +5869,98 @@ Int_t TCling::AutoParse(const char *cls)
+   static std::vector<std::string> int_and_libtable;
+
+   // Initialize the table
+   if (fFirstRun) {
+      clang::Preprocessor &PP = fInterpreter->getCI()->getPreprocessor();
+      const HeaderSearchOptions &HSOpts = PP.getHeaderSearchInfo().getHeaderSearchOpts();
+      auto ModulePaths(HSOpts.PrebuiltModulePaths);
+      std::sort(ModulePaths.begin(), ModulePaths.end());
+      auto last = std::unique(ModulePaths.begin(), ModulePaths.end());
+      ModulePaths.erase(last, ModulePaths.end());
+
+      cling::DynamicLibraryManager* dyLibManager = fInterpreter->getDynamicLibraryManager();
+
+      int cur_lib = 0;
+      for (auto Path : ModulePaths) {
+         StringRef DirPath(Path);
+         if (DirPath.empty() || !is_directory(DirPath) || Path == ".")
+            continue;
+
+         // Check if this is a build directory (which should contain pcm)
+         std::error_code EC;
+         bool is_build = false;
+         for (llvm::sys::fs::directory_iterator DirIt(DirPath, EC), DirEnd;
+               DirIt != DirEnd && !EC; DirIt.increment(EC)) {
+            std::string FileName(DirIt->path());
+            if (extension(FileName) == ".pcm") {
+               is_build = true;
+               break;
+            }
+         }
+
+         if (!is_build)
+            continue;
+
+         // Iterate over the directory
+         for (llvm::sys::fs::directory_iterator DirIt(DirPath, EC), DirEnd;
+               DirIt != DirEnd && !EC; DirIt.increment(EC)) {
+            std::string LibName(DirIt->path());
+            if (llvm::sys::fs::is_directory(LibName) || extension(LibName) != ".so")
+               continue;
+
+            // TCling::IsLoaded is incredibly slow!
+            if (dyLibManager->isLibraryLoaded(LibName.c_str()))
+               continue;
+
+            int_and_libtable.push_back(LibName);
+            auto SoFile = ObjectFile::createObjectFile(LibName);
+            if (SoFile) {
+               auto RealSoFile = SoFile.get().getBinary();
+               auto Symbols = RealSoFile->symbols();
+               for (auto S : Symbols) {
+                  // DO NOT load to table if symbol was weak or undefined
+                  if (S.getFlags() == SymbolRef::SF_Weak || S.getFlags() == SymbolRef::SF_Undefined) continue;
+
+                  std::string SymName = std::string(S.getName().get());
+                  // Put into map
+                  libtable.insert(std::make_pair(hash_fn(SymName), cur_lib));
+               }
+            }
+            cur_lib++;
+         }
+      }
+      fFirstRun = false;
+   }
+
+   auto key = libtable.find(hash_fn(mangled_name));
+   if (key != libtable.end()) {
+      std::string LoadedLibName = int_and_libtable[key->second];
+      if (!LoadedLibName.empty() && (gSystem->Load(LoadedLibName.c_str(), "", false) < 0))
+         std::cout << "Failed to load library" << std::endl;
+   }
+
+   void* addr = llvm::sys::DynamicLibrary::SearchForAddressOfSymbol(mangled_name.c_str());
```

Comment:
Can we not re-use the shared lib handle of the library we just opened? I know `TSystem::Load` doesn't give this, but we should be able to create a function that does!